### ISSUE TITLE
Makd: Don't generate a file importing all modules to create the unit tests

### ISSUE THREAD

**Issue by leandro-lucarella-sociomantic** [sociomantic-tsunami/makd#62]
This doesn't work properly with deprecated modules, as if the generated main imports the modules, the deprecations will be emitted anyway, which is not what we want.

We could just call `dmd` with all the same files we use to create the fake main.

----

@mihails-strasuns-sociomantic 

> Note though that it will require taking care of all dependencies manually though. It may be better course of actions to either enhance rdmd to support multiple file arguments or move automatic import file loading into dmd itself (and back-porting that to dmd1). Otherwise it will be duplicate code everywhere.

----

@mathias-baumann-sociomantic 

> Well, if you use the whole library you're bound to also have all required dependencies, right? Excluding external ones of course.

----

@mihails-strasuns-sociomantic 

> I mean that when building unit tests for module `a` imported modules from `b` library also need to be passed to command-line for plain dmd to work (this is what rdmd does internally for you)

----

@leandro-lucarella-sociomantic

> Yeah, good point. On the other hand, we can do a more smart handling of dependencies if we do it ourselves, and compile everything outside the repo without -unittest so we just compile and run the unittets for the particular project. Right now the unittest filtering is done at runtime but everything gets compiled anyway.

----

@mihails-strasuns-sociomantic 

> Why would one use rdmd then after re-implementing all its features in own build system? :)

----

@leandro-lucarella-sociomantic

> Who said we would? :trollface:
> 
> Seriously, the idea of doing another tool has been circling my mind for a while now. Truth is, `rdmd` is not a build tool, is a tool to easily allow run D programs as scripts that kind of provides a build tool since we had to do it anyway. For example, it writes to `/tmp` caching the results, and object files, which doesn't have any sense.
> 
> What we need keeps diverging from what `rdmd` provides. I'm not discarding either extending `rdmd` or to make a tool based on it, but as far as I seen in the `rdmd` code, there is a lot of complexity we didn't need and I'm not just it will be any easy to extend it. Maybe we can reuse some bits and pieces though.

----

@mihails-strasuns-sociomantic 

> This is a good point and the reason why I'd love to see some basic bits for dependency resolution moved to compiler itself. It will both result in faster clean compilation and reduce temptation to misuse `rdmd` as a build tool.

----

@leandro-lucarella-sociomantic

> Right now `dmd` does dependency resolution, and outputs it in some JSON format. This is what `rdmd` uses.

----

@mihails-strasuns-sociomantic 

> Instead DMD should automatically import the needed modules on its own, even if those were not supplied to the command-line. This will save one full semantic pass for all modules (`dmd -v` to get the module list + actual compilation) as it happens with `rdmd`.

----

@leandro-lucarella-sociomantic

> Then you need exclusion rules, because you don't want to include there either phobos, or the runtime, or tangort in our case, and then this doesn't work to have full control over how unittests are build either. I mean, I don't think it's completely a bad idea on itself, I just don't think it affect us a lot. We need something else.

----

@mihails-strasuns-sociomantic 

> Yes, of course. Pretty much all facilities rdmd has to make it work will need to be merged with compiler for such feature to make sense.
> 
> I really wonder if not compiling the tests is important enough to go that far. It is also worth reminding that in general D projects are guaranteed to compile only if all modules are built with same version flags (and -unittest introduces version flag). Otherwise there can be issue with symbol emitting, primarily for templates.

----

@leandro-lucarella-sociomantic

> Really? I never heard of that, I can understand it for `-debug` and `-version` options, and maybe even `-inline` or `-O`, but for `-unittest` it doesn't seem to have a lot of sense.
> 
> And yeah, for me it makes a lot of sense, for huge tests, compiling (and
running, unless you have a "hack" in the test runner) all the tests for
all the dependencies is overkill. I think is a needed feature if D wants
to be able to scale.

----

@mihails-strasuns-sociomantic 

> There is `version(unittest)` in D2 so it is the same as for any other version/debug flag. It is possible that if some template instance is referred from a module from inside the unittest version than other modules importing it will expect relevant symbol to be located in the matching object file and won't emit own symbol. And if it was compiled without `-unittest` but importing modules with it linking will fail.

----

@leandro-lucarella-sociomantic

> Mmm, sounds strange, template instances should be linked with the weak flag and emitted everywhere when doing separate compilation. I don't buy it yet :)

----

@mihails-strasuns-sociomantic 

> It was an optimization to speed up the linking added several releases ago : https://issues.dlang.org/show_bug.cgi?id=11284

----

@leandro-lucarella-sociomantic

> With `deprecated unittest` this is not a problem anymore, right?

----

@mihails-strasuns-sociomantic 

> No, the issue is still present and has no upstream solution. Currently I am manually filtering deprecated modules from testing to workaround it.
>
> Relevant dmd repo issue sociomantic/dmd#36

**Comment by mathias-lang-sociomantic** [comment]
Isn't that what `TEST_FILTER_OUT` is for ?

**Comment by leandro-lucarella-sociomantic** [comment]
Not exactly, is a general feature that serves as a hack to manually cope with this issue :)
